### ISSUE TITLE
Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true

### ISSUE THREAD

**Issue by coderextreme** [x3dom/x3dom#1282]
Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: <URL>
x3dom-full.debug.js:21721  Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently
x3dom.Texture.uploadTextMipmap @ x3dom-full.debug.js:21721
x3dom-full.debug.js:21721  Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently
x3dom.Texture.uploadTextMipmap @ x3dom-full.debug.js:21721
x3dom-full.debug.js:21721  Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently
x3dom.Texture.uploadTextMipmap @ x3dom-full.debug.js:21721
x3dom-full.debug.js:21721  Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently
x3dom.Texture.uploadTextMipmap @ x3dom-full.debug.js:21721
x3dom-full.debug.js:21721  Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently
x3dom.Texture.uploadTextMipmap @ x3dom-full.debug.js:21721
x3dom-full.debug.js:21721  Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently
x3dom.Texture.uploadTextMipmap @ x3dom-full.debug.js:21721
x3dom-full.debug.js:21721  Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently
x3dom.Texture.uploadTextMipmap @ x3dom-full.debug.js:21721
x3dom-full.debug.js:21721  Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently
x3dom.Texture.uploadTextMipmap @ x3dom-full.debug.js:21721
x3dom-full.debug.js:21721  Canvas2D: Multiple readback operations using getImageData are faster with the willReadFrequently attribute set to true. See: https://html.spec.whatwg.org/multipage/canvas.html#concept-canvas-will-read-frequently

**Comment by andreasplesch** [comment]
Thanks,

https://github.com/x3dom/x3dom/blob/434365a4bf0d52ddf1ee2d03c22d755bdbe64a5b/src/Texture.js#L875

seems the only place where `getImageData` is called, for setting up rendering Text. It is called rarely, eg. not every frame.

See https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext#willreadfrequently

Sounds like it trades off memory vs. performance. Probably not applicable here.

**Comment by coderextreme** [comment]
Joe's animation panel (choice of animation in as Texts) is not showing up
in x3dom.  I'll ask him to provide a copy to x3dom github.

Thanks, Joe.

**Comment by coderextreme** [comment]
Here's a picture of what's happening: X3DOM above, X_ITE below (the middle
one).

[image: image.png]

**Comment by andreasplesch** [comment]
Thanks. The picture did make it.

https://www.web3d.org/x3d/content/examples/HumanoidAnimation/Skin/BoxMan4AnimationPanelX3dom.xhtml

seems to work. Is the same panel ?

**Comment by coderextreme** [comment]
Yeah, works okay for me too.   Look at the console in Chrome or Edge.

John

**Comment by coderextreme** [comment]
What I mean with the picture is this example in X3DOM (on top) does not
match the menu below in X_ITE (below).  I will ask Joe if I can release the
model.

John

**Comment by andreasplesch** [comment]
The picture did not make it.

**Comment by coderextreme** [comment]
My understanding I’m still talking about the same picture I thought you
said did make it? A typo?

**Comment by andreasplesch** [comment]
Yes, it was a typo.
